### DISCUSSION TITLE
Dashboard can get access only when S3 is public

### DISCUSSION THREAD

**Discussion by coristo** [sorry-cypress/sorry-cypress#493] [Q&A]
## Summary

Sorry cypress doesn't connect to S3 bucket using IAM when bucket isn't public.

## How to reproduce

Try to connect local sorry cypress instance to S3 bucket.

## Environment

- sorry-cypress version: latest
- cypress version: latest
- platform: `docker`
- service: `director`

Hi, I've got locally Sorry cypress built using docker images. The problem is I can't get access to S3 bucket when it's not totally public. 
Is this the BUG or just IAM S3 Full Access isn't enough?

**Comment by agoldis** [comment]
@pawel-zajac90 the upload part uses a signed URL that should work fine with private s3 buckets, if it isn't working it's a bug.

The read part is problematic - if a browser tries to read private S3 URL it will fail, unless the URL is signed. The signing part might be implemented on the API service, that required more reading and exploration though. 

Any help is appreciated

**Comment by coristo** [comment]
@agoldis So for now there is no solution to allow users get public access to dashboard e.g. using nginx authorization system?
I cannot set up minio buckets externally so I decided to use AWS S3 buckets, but like I understand there is no possibility to let buckets be private?
When I will find some solution I will of course let you know. Just wanna know if there is no ready solution :)

EDIT:
What about using Dashboard on AWS?  Is this possible to send data using SSL?

**Comment by agoldis** [comment]
I was able to do that for https://currents.dev using [Cloudfront signing key](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/private-content-overview.html) but it required quite a lot of refactoring and also doesn't work with minio :(

**Comment by coristo** [comment]
@agoldis Actually upload doesn't work also...could you give me a hint where should I look for the issue?

**Comment by agoldis** [comment]
https://github.com/sorry-cypress/sorry-cypress/blob/0d1cd14a4b9fa8556f3977207783fdb3b1ef687e/packages/director/src/screenshots/s3/s3.ts#L43
That's the part that generates the upload URL

**Comment by coristo** [comment]
Okay @agoldis I did found what I needed. In my opinion In documentation should be an info, that we need to set AWS region (it was not so obviously for me) also.
I can't found this in minio docs. Is this possible to setup safe external access for it's buckets?
I'd like to use minio, not AWS buckets, because all data need to be on my servers. When I'm creating endpoints for dahsboard and directory it still cannot get access to bucket so cannot sned and read screenshots and videos.
My plan: set up external access to dashbaord using nginx user verification.
About "Any help is appreciated" I'm really interested to help with e.g kubernetes build if you'd like to. But looking for issues in code is not exactly my part of job so i need a little help to found the way.
 Sorry for my english, I can translate docs to Polish if you need ;)

**Comment by agoldis** [comment]
@pawel-zajac90 thanks for your feedback! Your English is quite fine - I am also not a native speaker, don't worry about it. The documentation is open-source, it's available at  https://github.com/sorry-cypress/gitbook - you can add information that you find useful.

Please take a look at https://github.com/sorry-cypress/charts - it is our Helm charts repo - any help with pull requests / issues would be great. 


Unfortunately I don't know how to setup minio privately :( it would some time to investigate. If you find a way to set up your system, please consider sharing in a blog post  / comment  - I will repost and share - that can help others!

**Comment by pitthecat** [comment]
> @agoldis Actually upload doesn't work also...could you give me a hint where should I look for the issue?

We are having the same problem. Our setup is 100% internal on AWS.

- Private VPC (Connected via VPN to company)
- DocumentDB for persistence 
- ECS Fargate service with API, Dashboard and Director (all latest images (v2.0.0-beta.16) from docker.io)
- Private S3 bucket (Block all public access)

Every upload of a video/screenshot fails to the S3 bucket. The ECS task role has full access to the bucket. As soon as I set the bucket to public every upload succeeds.

**Comment by agoldis** [comment]
@pitthecat 
Cypress agents upload screenshots directly to S3. If your bucket is completely private, then cypress agent won't be able to create new objects.

**Reply by pitthecat** [reply]
Thank you, that explains why it's not working. I decided to go with MinIO now. The MinIO bucket itself is "public", but in reality it's not really public because MinIO is running in our private AWS VPC.

**Comment by chrisdrobison** [comment]
We set the S3_ACL to private and uploads work just fine, but we've found with the dashboard that we can't view content because it assumes the bucket is publicly accessible. Is there a way for the dashboard to use signed URLs for showing the bucket content?

**Reply by agoldis** [reply]
https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/private-content-signed-urls.html

**Reply by chrisdrobison** [reply]
Yeah, we are familiar with that, but the sorry cypress dashboard isn't generating signed urls to display the content.

**Reply by agoldis** [reply]
It isn't today, but can be enhanced, PRs are welcome!

**Reply by chrisdrobison** [reply]
Yep, we are discussing doing that.
